docs/examples/word_embedding/word_embedding_training.md

```diff
@@ -0,0 +1,292 @@
+# Word Embeddings Training and Evaluation
+
+```python
+import warnings
+warnings.filterwarnings('ignore')
+
+import itertools
+import time
+import math
+import logging
+import random
+
+import mxnet as mx
+import gluonnlp as nlp
+import numpy as np
+from scipy import stats
+
+# context = mx.cpu()  # Enable this to run on CPU
+context = mx.gpu(0)  # Enable this to run on GPU
+```
+
+### Data
+Here we use the Text8 corpus from the [Large Text Compression
+Benchmark](http://mattmahoney.net/dc/textdata.html) which includes the first
+100
+MB of cleaned text from the English Wikipedia.
+
+```python
+dataset = nlp.data.Text8()
+print('# sentences:', len(dataset))
+for sentence in dataset[:3]:
+    print('# tokens:', len(sentence), sentence[:5])
+```
+
+We pass the dataset to the `prepare_batches` function from the `utils.py` of
+this notebook.
+
+The following actions are taken:
+- subsampling
+- negative
+samples
+- subwords in CSR
+
+Some words such as "the", "a", and "in" are very
+frequent.
+One important trick applied when training word embeddings is to
+subsample the dataset
+according to the token frequencies. [1] proposes to
+discard individual
+occurences of words from the dataset with probability
+$$P(w_i) = 1 - \sqrt{\frac{t}{f(w_i)}}$$
+
+where $f(w_i)$ is the frequency with
+which a word is
+observed in a dataset and $t$ is a subsampling constant
+typically chosen around
+$10^{-5}$.
+
+[1] Mikolov, Tomas, et al. “Distributed
+representations of words and phrases and their compositionality.” Advances in
+neural information processing systems. 2013.
+
+```python
+import utils
+(batches, vocab, subword_function) = utils.prepare_batches(dataset, ngrams=[3,4,5,6],
+                                                           num_subwords=10000, num_negatives=5, batch_size=16384, window=5)
+```
+
+## Training word embeddings with subword information
+
+`gluonnlp` provides the
+concept of a SubwordFunction which maps words to a list of indices representing
+their subword.
+Possible SubwordFunctions include mapping a word to the sequence
+of it's characters/bytes or hashes of all its ngrams.
+
+FastText models use a
+hash function to map each ngram of a word to a number in range `[0,
+num_subwords)`. We include the same hash function.
+
+### Concept of a
+SubwordFunction
+
+```python
```

Comment:
This would become a regular python code block without being run or rendered. Follow the following guide to correct this: http://gluon-nlp.mxnet.io/how_to/contribute.html#contribute-examples